### PR TITLE
feat: add NAFNet Implementation

### PR DESCRIPTION
**Pull request by soumik12345**
Fixes https://github.com/soumik12345/restorers/issues/24

### CODE REVIEW

**Comment by Samarendra109** [comment]
@SauravMaheshkar could you please review this PR? Thanks.

File: restorers/model/nafnet/nafblock.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        super().__init__(**kwargs)
+        self.factor = factor
+
+    def call(self, x: tf.Tensor, *args, **kwargs) -> tf.Tensor:
+        x = tf.expand_dims(x, axis=-1)
+        return tf.reduce_prod(
+            tf.concat(tf.split(x, num_or_size_splits=self.factor, axis=-2), axis=-1),
+            axis=-1,
+        )
+
```
Comment (SauravMaheshkar) [inline restorers/model/nafnet/nafblock.py:25]: Could you add a `get_config` fn to all custom layers ? It helps with distributed training in some instances. [Reference Implementation](https://github.com/SauravMaheshkar/Knowledge-Distillation/blob/ef5b92bd53852df5d920746ba19d870360467e5b/src/backend/tensorflow/modules/affine.py#L46)

File: restorers/model/nafnet/nafnet.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    Wrapper Class for tf.nn.depth_to_space
+    """
+
+    def __init__(self, upscale_factor: int, **kwargs) -> None:
+        super().__init__(**kwargs)
+        self.upscale_factor = upscale_factor
+
+    def call(self, inputs: tf.Tensor, *args, **kwargs) -> tf.Tensor:
+        return tf.nn.depth_to_space(inputs, self.upscale_factor)
+
```
Comment (SauravMaheshkar) [inline restorers/model/nafnet/nafnet.py:21]: Could you add a `get_config` fn to all custom layers ? It helps with distributed training in some instances. [Reference Implementation](https://github.com/SauravMaheshkar/Knowledge-Distillation/blob/ef5b92bd53852df5d920746ba19d870360467e5b/src/backend/tensorflow/modules/affine.py#L46)

File: restorers/model/nafnet/nafnet.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
```
Comment (SauravMaheshkar) [inline restorers/model/nafnet/nafnet.py:22]: Could you add a `get_config` fn to all custom models as well ? It helps with distributed training in some instances. [Reference Implementation](https://github.com/SauravMaheshkar/Knowledge-Distillation/blob/ef5b92bd53852df5d920746ba19d870360467e5b/src/backend/tensorflow/modules/affine.py#L46)

**Review by SauravMaheshkar — CHANGES_REQUESTED** [review]
Looks good overall, just need to add some `get_config` functions

File: restorers/model/nafnet/nafblock.py (hunk 1 above)
Comment (Samarendra109) [reply restorers/model/nafnet/nafblock.py:25]: Could you please explain a bit, what get_config does? Like what class members am I supposed to add to get_config? Is it just the constructor arguments ?

**Comment by Samarendra109** [comment]
@SauravMaheshkar configs are added. Have a final look. I think we can merge.

**Review by SauravMaheshkar — APPROVED** [review]
LGTM ❤️

**Pull request merged**